### ISSUE TITLE
Windows implementation of UnmapOrDie fails to free address space

### ISSUE THREAD

**Issue by rnk** [google/sanitizers#658]
This bug has been present since Timur added the first call to VirtualFree in r151366. Fixing it will not be trivial. See the documentation for VirtualFree:
https://msdn.microsoft.com/en-us/library/windows/desktop/aa366892(v=vs.85).aspx
We should be passing MEM_RELEASE instead of MEM_DECOMMIT to return the virtual address space to the pool. However, Windows does not support unmapping part of a virtual address space allocation:

> MEM_RELEASE
> Releases the specified region of pages. After this operation, the pages are in the free state.
> If you specify this value, dwSize must be 0 (zero), and lpAddress must point to the base address returned by the VirtualAlloc function when the region is reserved. The function fails if either of these conditions is not met.

Basically, VirtualFree cannot directly implement the aspect of munmap that returns part of an allocation to the OS. This might not sound like a big deal, but MmapAlignedOrDie is implemented as mmap(size + alignment) followed by two munmaps to return the unused pages. Since our Windows implementation of munmap doesn't actually return the pages to the OS, WinASan is spending twice as much address space as it should for heap allocations. :(

We may be able to fix MmapAlignedOrDie by doing something like:
```
void *p = VirtualAlloc(0, size + alignment);
void *q = align(p, alignment);
VirtualFree(p, 0, MEM_DECOMMIT);
return VirtualAlloc(q, size);
```

**Comment by eugenis** [comment]
I think we also have this optimization when unpoisoning of a large region is implemented by partially unmapping / remapping the corresponding shadow pages. This can be simply disabled if nothing else.

**Comment by morehouse** [comment]
Any progress on this?
